File: Task 4/Java/logistics/src/main/java/com/logistics/Components/Stock.java
package com.logistics.Components;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import com.logistics.Components.Product.Product;
import com.logistics.Serialization.PBLogistics.PBStock;
import com.logistics.Serialization.PBLogistics.PBStockItem;

public class Stock {
    public static class stockItem_t {
        public Product p;
        public int q;
        public stockItem_t(){this.p = null; this.q = 0;}
        public stockItem_t(Product p, int q){this.p = p; this.q = q;}
    }

    private List<stockItem_t> stockList;

    // Default constructor
    public Stock() {
        stockList = new ArrayList<>();
    }

    // Getter for the stock list
    public List<stockItem_t> getStockList() {return stockList;}

    // Adds a stock item to the stock list
    public void addStock(stockItem_t newStock) {
        if (newStock.p == null) {
            return;
        }

        // Check for duplication of item
        if (getStock(newStock.p.getName()) == null) {
            stockList.add(newStock);
            return;
        }

        System.out.println("ERROR: Item with the same name is already in stock");
    }

    // Removes a stock item from the stock list
    public void removeStock(String stockName) {
        // Search for the item to remove
        for (int i = 0; i < stockList.size(); i++) {
            if (stockList.get(i).p.getName().equals(stockName)) {
                stockList.remove(i);
                System.out.println("\nDeleted Successfully\n");
                return;
            }
        }
        System.out.println("\nERROR: Stock not found"); // not deleted
    }

    // Updates a stock item in the stock list
    public void updateStock(String stockName, stockItem_t newStock) {
        if (newStock.p == null) {
            return;
        }
        stockItem_t toUpdate = getStock(stockName);
        if (toUpdate != null) {
            // Check that the product either has the same name as the one being updated or a different name
            if (newStock.p.getName().equals(toUpdate.p.getName()) ||
               (!newStock.p.getName().equals(toUpdate.p.getName()) && getStock(newStock.p.getName()) == null)) {
                removeStock(stockName);
                addStock(newStock);
                System.out.println("\nStock Updated\n");
                return;
            }
            System.out.println("Stock was NOT Updated.\n");
        } else {
            System.out.println("Stock not found.\n");
        }
    }

    // Increases the quantity of a stock item in the stock list
    public void increaseStock(String stockName, int newQuantity) {
        // Search for the item to increase stock
        stockItem_t toIncrease = getStock(stockName);
        if (toIncrease != null) {
            toIncrease.q += newQuantity;
            return;
        }
        System.out.println("Stock not found.\n");
    }

    // Clears the stock list and manages memory
    public void clearStock() {
        // Memory Management
        for (stockItem_t item : stockList) {
            item.p = null;
        }
        stockList.clear();
    }

    // Retrieves a stock item from the stock list based on its name
    public stockItem_t getStock(String stockName) {
        for (stockItem_t stock : stockList) {
            if (stock.p.getName().equals(stockName)) {
                return stock;
            }
        }
        return null;
    }

    // Displays all stock items
    public void displayStock() {
        // Traverse and display
        System.out.println("Stock Items:");
        for (stockItem_t item : stockList) {
            System.out.printf("\n%s", item.p.display_product_info());
            System.out.printf("\nQuantity: %s\n", item.q);
        }
    }

    // Displays a specific stock item based on its name
    public void displayStock(String stockName) {
        // Traverse and display
        stockItem_t item = getStock(stockName);
        if (item != null) {
            System.out.printf("\n%s", item.p.display_product_info());
            System.out.printf("\nQuantity: %s\n", item.q);
            return;
        }
        System.out.println("\nERROR: Stock not found");
    }





    // ========================================== PROTOBUFFERS


    public void copyToProto(PBStock.Builder protobufStockList){
        protobufStockList.addAllStock(
            stockList.stream().map(stockItem -> {
                PBStockItem.Builder message = PBStockItem.newBuilder();
                stockItem.p.copyToProto(message);
                message.setQuantity(stockItem.q);
                return message.build();
            }).collect(Collectors.toList())
        );
    }


   public void copyFromProto(PBStock protobufStockList) {
        clearStock(); // Clear the existing stock

        // Iterate through protobuf stockitems and create corresponding Java StockItems objects
        for (PBStockItem message : protobufStockList.getStockList()) {

            if (message.hasAccessory()) {               // -------------------------------- ACCESSORY

                Product p = ProductFactory.createProduct("Accessory");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));

            } else if (message.hasClothing()) {         // -------------------------------- CLOTHING

                Product p = ProductFactory.createProduct("Clothing");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));

            } else if (message.hasFood()) {             // -------------------------------- FOOD

                Product p = ProductFactory.createProduct("Food");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasDrink()) {            // -------------------------------- DRINK

                Product p = ProductFactory.createProduct("Drink");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasTech()) {              // -------------------------------- TECH

                Product p = ProductFactory.createProduct("Tech");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasAppliance()) {        // -------------------------------- APPLIANCE

                Product p = ProductFactory.createProduct("Appliance");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasDecor()) {             // -------------------------------- DECOR

                Product p = ProductFactory.createProduct("Decor");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasStorage()) {           // -------------------------------- STORAGE

                Product p = ProductFactory.createProduct("Storage");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasDigital()) {           // -------------------------------- DIGITAL

                Product p = ProductFactory.createProduct("Digital");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasPhysical()) {         // --------------------------------  PHYSICAL

                Product p = ProductFactory.createProduct("Physical");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else if (message.hasPlants()) {           // -------------------------------- PLANTS

                Product p = ProductFactory.createProduct("Plants");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            }  else if (message.hasSeeds()) {           // -------------------------------- SEEDS

                Product p = ProductFactory.createProduct("Seeds");
                p.copyFromProto(message);
                addStock(new stockItem_t(p, message.getQuantity()));
                
            } else{
                System.out.println("ERROR: Loaded Invalid Item");
            }
        }
    }
}
     

File: Task 3/C++Stock Viewer/Stock.cpp
#include "Stock.h"

Stock::Stock(){}

void Stock::addStock(Stock::stockItem_t* NewStock){
    if(NewStock->p == nullptr){return;};

    // Check For duplication of item
    if(getStock(NewStock->p->getName()) == nullptr){
        stockList.push_back(NewStock);
        return;
    }

    printf("ERROR: Item with Same name is already in stock\n");
}

void Stock::removeStock(const std::string stockName){
    // Search for item to remove
    for(int i = 0; i<stockList.size(); i++){
        if(stockList[i]->p->getName() == stockName){
            delete stockList[i]->p;
            delete stockList[i];
            stockList.erase(stockList.begin()+i);
            printf("\nDeleted Successfully\n");
            return;
        }
    }
    printf("\nERROR: Stock not found\n"); // not deleted
}

void Stock::updateStock(const std::string stockName, Stock::stockItem_t* NewStock){
    if(NewStock->p == nullptr){return;}
    Stock::stockItem_t* toUpdate = getStock(stockName); 
    if(toUpdate != nullptr){

         // Check that Transport either has the same code as the one being updated or a different code
        if ((NewStock->p->getName() == toUpdate->p->getName()) ||
            (NewStock->p->getName() != toUpdate->p->getName() && getStock(NewStock->p->getName()) == nullptr)) {
            removeStock(stockName);
            addStock(NewStock);
            printf("\nStock Updated\n");
            return;
        }
        printf("Stock was NOT Updated.\n");
    } else {
        printf("Stock not found.\n");
    }
}

void Stock::increaseStock(const std::string stockName, int newQuantity){
    
    // Search for item to increase stock
    Stock::stockItem_t* toIncrease = getStock(stockName); 
    if(toIncrease != nullptr){
        toIncrease->q += newQuantity;
        return;
    }
    printf("Stock not found.\n");
}

void Stock::clearStock(){
    // Memory Management
    for (Stock::stockItem_t* item : stockList) {
        delete item->p;
        delete item;
    }
    stockList.clear();
}

Stock::stockItem_t* Stock::getStock(const std::string stockName){
    for(stockItem_t* stock : stockList){
        if(stock->p->getName() == stockName){
            return stock;
        }
    }
    return nullptr;
}

void Stock::displayStock(){
    // Traverse and display
    std::cout << "Stock Items:\n";
    for (Stock::stockItem_t* item : stockList) {
        printf("\n%s", item->p->display_product_info().c_str());
        printf("\nQuantity: %s\n", std::to_string(item->q).c_str());
    }
}
void Stock::displayStock(const std::string stockName){
    // Traverse and display
    Stock::stockItem_t* item = getStock(stockName);
    if(item != nullptr){
        printf("\n%s", item->p->display_product_info().c_str());
        printf("\nQuantity: %s\n", std::to_string(item->q).c_str());
        return;
    }
    printf("\nERROR: Stock not found\n");
}



// ========================================== PROTOBUFFERS

void Stock::copyToProto(PBStock protobufStockList) {
    for (stockItem_t* stock : stockList) {
        PBStockItem* message = protobufStockList.add_stock();
        stock->p->copyToProto(message);
        message->set_quantity(stock->q);
    }
}

void Stock::copyFromProto(PBStock protobufStockList) {
    clearStock(); // Clear the existing stock

    // Iterate through protobuf stock items and create corresponding Java StockItems objects
    for (const auto& message : protobufStockList.stock()) {
        if (message.has_accessory()) {               // -------------------------------- ACCESSORY
            Accessory* accessory = new Accessory();
            accessory->copyFromProto(message);
            addStock(new stockItem_t{accessory, message.quantity()});
        } else if (message.has_clothing()) {         // -------------------------------- CLOTHING
            Clothing* clothing = new Clothing();
            clothing->copyFromProto(message);
            addStock(new stockItem_t{clothing, message.quantity()});
        } else if (message.has_food()) {             // -------------------------------- FOOD
            Food* food = new Food();
            food->copyFromProto(message);
            addStock(new stockItem_t{food, message.quantity()});
        } else if (message.has_drink()) {            // -------------------------------- DRINK
            Drink* drink = new Drink();
            drink->copyFromProto(message);
            addStock(new stockItem_t{drink, message.quantity()});
        } else if (message.has_tech()) {              // -------------------------------- TECH
            Tech* tech = new Tech();
            tech->copyFromProto(message);
            addStock(new stockItem_t{tech, message.quantity()});
        } else if (message.has_appliance()) {        // -------------------------------- APPLIANCE
            Appliance* appliance = new Appliance();
            appliance->copyFromProto(message);
            addStock(new stockItem_t{appliance, message.quantity()});
        } else if (message.has_decor()) {             // -------------------------------- DECOR
            Decor* decor = new Decor();
            decor->copyFromProto(message);
            addStock(new stockItem_t{decor, message.quantity()});
        } else if (message.has_storage()) {           // -------------------------------- STORAGE
            Storage* storage = new Storage();
            storage->copyFromProto(message);
            addStock(new stockItem_t{storage, message.quantity()});
        } else if (message.has_digital()) {           // -------------------------------- DIGITAL
            Digital* digital = new Digital();
            digital->copyFromProto(message);
            addStock(new stockItem_t{digital, message.quantity()});
        } else if (message.has_physical()) {         // --------------------------------  PHYSICAL
            Physical* physical = new Physical();
            physical->copyFromProto(message);
            addStock(new stockItem_t{physical, message.quantity()});
        } else if (message.has_plants()) {           // -------------------------------- PLANTS
            Plants* plants = new Plants();
            plants->copyFromProto(message);
            addStock(new stockItem_t{plants, message.quantity()});
        } else if (message.has_seeds()) {           // -------------------------------- SEEDS
            Seeds* seeds = new Seeds();
            seeds->copyFromProto(message);
            addStock(new stockItem_t{seeds, message.quantity()});
        } else {
            std::cout << "ERROR: Loaded Invalid Item" << std::endl;
        }
    }
}

